src/solver/brent/mod.rs

```diff
@@ -0,0 +1,181 @@
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Brent's method
+//!
+//! A root-finding algorithm combining the bisection method, the secant method
+//! and inverse quadratic interpolation. It has the reliability of bisection
+//! but it can be as quick as some of the less-reliable methods.
+//!
+//! # References:
+//!
+//! https://en.wikipedia.org/wiki/Brent%27s_method
+//!
+
+/// Implementation of Brent's optimization method,
+/// see https://en.wikipedia.org/wiki/Brent%27s_method
+use crate::prelude::*;
+use serde::{Deserialize, Serialize};
+use std::f64;
+use std::fmt;
+
+/// Error to be thrown if Brent is initialized with improper parameters.
+#[derive(Debug)]
+pub struct BrentError;
+
+impl fmt::Display for BrentError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Brent error: f(min) and f(max) must have different signs."
+        )
+    }
+}
+
+impl std::error::Error for BrentError {
+    fn description(&self) -> &str {
+        "Brent requires a min and max value bracketing the root of the function."
+    }
+}
+
+/// Brent's method
+///
+/// A root-finding algorithm combining the bisection method, the secant method
+/// and inverse quadratic interpolation. It has the reliability of bisection
+/// but it can be as quick as some of the less-reliable methods.
+///
+/// # References:
+/// https://en.wikipedia.org/wiki/Brent%27s_method
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Brent {
+    tol: f64,
+    a: f64,
+    b: f64,
+    c: f64,
+    d: f64,
+    e: f64,
+    fa: f64,
+    fb: f64,
+    fc: f64,
+}
+
+impl Brent {
+    /// Constructor
+    /// The values `min` and `max` must bracketing the root of the function.
+    /// The parameter `tol` specifies the relative error to be targeted.
+    pub fn new(min: f64, max: f64, tol: f64) -> Brent {
+        Brent {
+            tol: tol,
+            a: min,
+            b: max,
+            c: max,
+            d: f64::NAN,
+            e: f64::NAN,
+            fa: f64::NAN,
+            fb: f64::NAN,
+            fc: f64::NAN,
+        }
+    }
+}
+
+impl<O> Solver<O> for Brent
+where
+    O: ArgminOp<Param = f64, Output = f64>,
+{
+    const NAME: &'static str = "Brent";
+
+    fn init(
+        &mut self,
+        op: &mut OpWrapper<O>,
+        // Brent maintains its own state
+        _state: &IterState<O>,
+    ) -> Result<Option<ArgminIterData<O>>, Error> {
+        self.fa = op.apply(&self.a)?;
+        self.fb = op.apply(&self.b)?;
+        if self.fa * self.fb > 0.0 {
+            return Err(Error::from_boxed_compat(Box::new(BrentError)));
+        }
+        self.fc = self.fb;
+        Ok(Some(
+            ArgminIterData::new().param(self.b).cost(self.fb.abs()),
+        ))
+    }
+
+    fn next_iter(
+        &mut self,
+        op: &mut OpWrapper<O>,
+        // Brent maintains its own state
```

Comment:
`IterState` will automatically track your current parameter vector and the current best parameter vector and their corresponding cost function values. Maybe you could reduce some of the state of the `Brent` struct with this (if you want, I personally think this is a matter of taste).

Reply:
I considered that, but the fields `prev_param`, `prev_best_param`, etc. doesn't really fit the meaning and it seemed to be an abuse to use them for other purpose. I also wasn't sure if these values might be set automatically by the the `Executor`, in which case it wouldn't work anyway.

Reply:
I've made similar experiences with some solvers. The `IterState` design works for most solvers and adds a lot of convenience, but for some solvers it is just not appropriate.